jOOQ/pom.xml

```diff
@@ -101,6 +101,11 @@
             <scope>provided</scope>
             <optional>true</optional>
         </dependency>
+        <dependency>
+            <groupId>org.jetbrains.kotlin</groupId>
+            <artifactId>kotlin-reflect</artifactId>
+            <version>1.2.30</version>
+        </dependency>
```

Comment:
Hmm, we cannot actually have this dependency in jOOQ, not even as `<scope>provided</scope>`. What I meant by "meta-reflecting" is that we need to reflect on this reflection API. I'll comment again in the actual code.

Reply:
I see. I noticed there are almost no dependencies. I will try to work around it . To satisfy my curiosity, why can't jOOQ have that dependency?

Reply:
It is a very very very *very* ***VERY*** bad idea for a library to have dependencies. Blog post here:
https://blog.jooq.org/2016/08/11/all-libraries-should-follow-a-zero-dependency-policy

Reply:
This is what I propose as a solution:

Kotlin encodes metadata in a specific protobuf format. `kotlin-reflect` parses that and provides API to access all of the metadata. That approach was used in this PR. Since we can't depend on `kotlin-reflect` we would have to get the `Metadata` annotation using Java reflection and parse the protobuf data ourselves.

There is an official project `kotlinx.reflect.lite` which is a very lightweight implementation of `kotlin-reflect` which only supports reading parameter names and nullability. After reading the article you wrote I think you could accept that we rip out only the most necessary parts of this library to make extracting parameter names from the `Metadata` annotation possible. I would except this could consist of only one additional class being added to jOOQ. The `kotlinx.reflect.lite` library is licensed as Apache License 2.0 so it could be included in both the free and commercial version of jOOQ.

Here is a minimal example of obtaining the parameter names using this library:

```java
ClassMetadata classMetadata = ReflectionLite.INSTANCE.loadClassMetadata(PersonKotlin.class);
ConstructorMetadata constructorMetadata = classMetadata.getConstructor(p.getClass().getDeclaredConstructors()[0]);
List<ParameterMetadata> parameters = constructorMetadata.getParameters();
List<String> names = new ArrayList<>();
for (ParameterMetadata parameter : parameters) {
    names.add(parameter.getName());
}
System.out.println(names);
```

Let me know what you think.

Reply:
It seems that the `kotlinx.reflect.lite` depends on `com.google.protobuf`. I don't see how we could get rid of that. It makes very little sense to reimplement protobuf parser from scratch.

Reply:
Of course not. As I said, we can have an implicit runtime dependency on `kotlin-reflect` by meta-reflecting on its API. If it's on the classpath, we can use it. That's what we're doing with vendor-specific JDBC API as well, including ojdbc.

E.g. instead of 

```java
KClass<? extends E> kotlinClass = JvmClassMappingKt.getKotlinClass(type);
```

Write

```java
Reflect.on("com.example.JvmClassMappingKt").call("getKotlinClass", type);
```

Reply:
Oh I see. I did not realize this could be done. I will look into it.